code/modules/unit_tests/unit_test.dm

```diff
@@ -0,0 +1,162 @@
+/*
+
+Usage:
+Override /Run() to run your test code
+
+Call Fail() to fail the test (You should specify a reason)
+
+You may use /New() and /Destroy() for setup/teardown respectively
+
+You can use the run_loc_floor_bottom_left and run_loc_floor_top_right to get turfs for testing
+
+Based on Unit Test Framework by: Cyberboss
+
+*/
+
+
+#define LANDMARK_BOTTOM_LEFT "bottom_left"
+#define LANDMARK_TOP_RIGHT "top_right"
+
+var/global/datum/unit_test_controller/unit_tests = new()
+/datum/unit_test_controller
+	var/datum/unit_test/current_test
+	var/failed_any_test = FALSE
+	var/test_log
+	var/dmm_suite/reservation
+
+/datum/unit_test_controller/proc/log_test(text)
+	if(src.test_log)
+		src.test_log << "\[[time2text(world.timeofday,"hh:mm:ss")]]: [text]"
+	world.log << text
+
+/datum/unit_test_controller/proc/run_tests()
+	if (isnull(reservation))
+		var/loaded = file2text('code/modules/unit_tests/unit_tests.dmm')
+		reservation = new/dmm_suite()
+		reservation.read_map(loaded, 1, 1, 1,'code/modules/unit_tests/unit_tests.dmm', DMM_OVERWRITE_MOBS | DMM_OVERWRITE_OBJS)
+		sleep(0)
+	LAGCHECK(LAG_HIGH)
+
+	var/tests_to_run = childrentypesof(/datum/unit_test)
+	for (var/_test_to_run in tests_to_run)
+		var/datum/unit_test/test_to_run = _test_to_run
+		if (initial(test_to_run.focus))
+			tests_to_run = list(test_to_run)
+			break
+
+	var/list/test_results = list()
+
+	for(var/I in tests_to_run)
+		var/datum/unit_test/test = new I
+
+		src.current_test = test
+		var/duration = TIME
+
+		test.Run()
+
+		duration = TIME - duration
+		src.current_test = null
+		src.failed_any_test |= !test.succeeded
+
+		var/list/log_entry = list("[test.succeeded ? "PASS" : "FAIL"]: [I] [duration / 10]s")
+		var/list/fail_reasons = test.fail_reasons
+
+		for(var/J in 1 to length(fail_reasons))
+			log_entry += "\tREASON #[J]: [fail_reasons[J]]"
+		var/message = log_entry.Join("\n")
+		log_test(message)
+
+		test_results[I] = list("status" = test.succeeded ? UNIT_TEST_PASSED : UNIT_TEST_FAILED, "message" = message, "name" = I)
+
+		qdel(test)
+
+		LAGCHECK(LAG_HIGH)
+
+	var/file_name = "data/unit_tests.json"
+	fdel(file_name)
+	file(file_name) << json_encode(test_results)
+
+	//Fail Automaton
+	if(src.failed_any_test)
+		for(var/test in test_results)
+			if(test_results[test]["status"]==UNIT_TEST_FAILED)
+				text2file("[test_results[test]["message"]]", "errors.log")
+
+	//We done, lets bail when mapSwitcher awakens
+	while(!mapSwitcher)
+		sleep(1)
```

Comment:
1 what

Reply:
Context of human time isn't meaningful here. Ticks

Reply:
🤔
![image](https://user-images.githubusercontent.com/4741640/129636194-db428696-2b2b-4a7f-9ed8-b0a64a2b1d62.png)

Reply:
That makes sense, there would be no real reason to couple those items.  That would only de-feature sleep.